src/node_file.cc

```diff
@@ -966,32 +967,67 @@ void Access(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
-void Close(const FunctionCallbackInfo<Value>& args) {
+static void Close(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
 
-  const int argc = args.Length();
-  CHECK_GE(argc, 1);
+  CHECK_EQ(args.Length(), 2);  // fd, req
 
   int fd;
   if (!GetValidatedFd(env, args[0]).To(&fd)) {
     return;
   }
   env->RemoveUnmanagedFd(fd);
 
-  if (argc > 1) {  // close(fd, req)
-    FSReqBase* req_wrap_async = GetReqWrap(args, 1);
-    CHECK_NOT_NULL(req_wrap_async);
-    FS_ASYNC_TRACE_BEGIN0(UV_FS_CLOSE, req_wrap_async)
-    AsyncCall(env, req_wrap_async, args, "close", UTF8, AfterNoArgs,
-              uv_fs_close, fd);
-  } else {  // close(fd)
-    FSReqWrapSync req_wrap_sync("close");
-    FS_SYNC_TRACE_BEGIN(close);
-    SyncCallAndThrowOnError(env, &req_wrap_sync, uv_fs_close, fd);
-    FS_SYNC_TRACE_END(close);
+  FSReqBase* req_wrap_async = GetReqWrap(args, 1);
+  CHECK_NOT_NULL(req_wrap_async);
+  FS_ASYNC_TRACE_BEGIN0(UV_FS_CLOSE, req_wrap_async)
+  AsyncCall(
+      env, req_wrap_async, args, "close", UTF8, AfterNoArgs, uv_fs_close, fd);
+}
+
+// Separate implementations are required to provide fast API for closeSync.
+// If both close and closeSync are implemented using the same function, and
+// if a fast API implementation is added for closeSync, close(fd, req) will
+// also trigger the fast API implementation and cause an incident.
+// Ref: https://github.com/nodejs/node/issues/53902
+static void CloseSync(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_EQ(args.Length(), 1);
+
+  int fd;
+  if (!GetValidatedFd(env, args[0]).To(&fd)) {
+    return;
+  }
+  env->RemoveUnmanagedFd(fd);
+  FSReqWrapSync req_wrap_sync("close");
+  FS_SYNC_TRACE_BEGIN(close);
+  SyncCallAndThrowOnError(env, &req_wrap_sync, uv_fs_close, fd);
+  FS_SYNC_TRACE_END(close);
+}
+
+static void FastCloseSync(Local<Object> recv,
+                          const uint32_t fd,
+                          // NOLINTNEXTLINE(runtime/references) This is V8 api.
+                          v8::FastApiCallbackOptions& options) {
+  Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
```

Comment:
not 100% sure but i think you may need to declare a HandleScope using `options.isolate` prior to opening this context handle, otherwise its leaking into whatever ambient HandleScope is present up the stack.

Reply:
We don't use the latest V8 version to leverage `options.isolate`. Node 22 uses `12.4.254.21-node.15`. Without `options.isolate` change landed on v8, I don't think there's much we can do.

Reply:
> Env is reused across all contexts in an Isolate right? if that's the case, maybe we can just store an env pointer on the isolate instead of reaching through the context?

That is not strictly maintained, though we try to keep it that way (e.g. see discussions in https://github.com/nodejs/node/pull/47855).

I think we discussed using `Isolate::SetData()` somewhere else, though the issue is that there are only 4 slots, and it may be a problem for Node.js embedders that also are already using those slots.

Reply:
Re. handle scopes, you can already get the isolate from the context and create a handle scope from that. I am not sure if that's really an issue, though - we almost never create handle scopes in our callbacks, it seems fine to assume that fast APIs work in the same way?

Reply:
> I think we discussed using `Isolate::SetData()` somewhere else, though the issue is that there are only 4 slots, and it may be a problem for Node.js embedders that also are already using those slots.

We used to store `IsolateData*` there in the past: c3cd453cbae84ff6d639a2b8d7776a15b3e3431b

That's not quite `Environment*` (which, yeah, technically doesn't always need to be the same for a single `Isolate*` – if we ever wanted to productionize something like [synchronous-worker](https://github.com/addaleax/synchronous-worker), we could).

I agree that that's probably not ideal for embedders, and we should avoid re-entangling these bits if we can. If we need to pass something like an `Environment*` through, would something like a `v8::External*` or `BaseObject*` work?

Reply:
Actually I am hitting the handle scope issue in https://github.com/nodejs/node/pull/54384 so yes it is required - I used a fix with v8::Isolate::TryGetCurrent there since v22 is stuck with 12.4, on main which has 12.8 this can use the isolate from fallback options to create the handle scope.